File: langs/python/tests/unit/helpers/test_dispatcher.py
"""IDispatcher-equivalent backed by deterministic Rx test schedulers."""

from __future__ import annotations

from reactivex.abc import SchedulerBase
from reactivex.testing import TestScheduler


class TestDispatcher:
    """Dispatcher with TestScheduler foreground + background for deterministic time.

    Satisfies the ``Dispatcher`` Protocol structurally via ``.foreground`` /
    ``.background`` properties; does not inherit.
    """

    __test__ = False  # tell pytest this is not a test class

    def __init__(self) -> None:
        self.foreground_scheduler: TestScheduler = TestScheduler()
        self.background_scheduler: TestScheduler = TestScheduler()

    @property
    def foreground(self) -> SchedulerBase:
        return self.foreground_scheduler

    @property
    def background(self) -> SchedulerBase:
        return self.background_scheduler

    def advance_all(self, ticks: int = 1) -> None:
        self.foreground_scheduler.advance_by(ticks)
        self.background_scheduler.advance_by(ticks)

File: langs/csharp/tests/VMx.Tests/Helpers/TestDispatcher.cs
using System.Reactive.Concurrency;
using Microsoft.Reactive.Testing;
using VMx.Services;

namespace VMx.Tests.Helpers;

/// <summary>
/// IDispatcher backed by deterministic Rx TestSchedulers so tests can
/// advance virtual time precisely. Foreground and Background each get
/// their own scheduler so cross-scheduler dispatch is observable.
/// </summary>
public sealed class TestDispatcher : IDispatcher
{
    public TestScheduler ForegroundScheduler { get; } = new();
    public TestScheduler BackgroundScheduler { get; } = new();

    public IScheduler Foreground => ForegroundScheduler;
    public IScheduler Background => BackgroundScheduler;

    public void AdvanceAll(long ticks = 1)
    {
        ForegroundScheduler.AdvanceBy(ticks);
        BackgroundScheduler.AdvanceBy(ticks);
    }
}
